README.md
# BookerPdfTool

iBooker/DracoUnion 知识库 PDF 处理工具

一套面向电子书 / 扫描版 PDF 的命令行工具，提供压缩、抽取图片、转 HTML、黑白反转纠正、AI 超分增强、打包成 PDF、扫描版甄别、PDF 去重、加密压缩包破解、PDG 转图片、Office 转 PDF 等能力。

## 安装

通过 pip（推荐）：

```
pip install BookerPdfTool
```

从源码安装：

```
pip install git+https://github.com/DracoUnion/BookerPdfTool.git
```

Python 版本要求：>= 3.6。

安装后可通过 `BookerPdfTool` 或 `pdf-tool` 两个命令调用（两者等价），也可以使用 `python -m BookerPdfTool`。

### 可选依赖

+   `anime4k-auto` / `auto` 命令需要 [Anime4KCPP_CLI](https://github.com/TianZerL/Anime4KCPP)，并将其所在目录加入系统 PATH。
+   `fm-office` 命令仅在 Windows 上可用（基于 Office COM 组件）。
+   `crack-zip` 破解 RAR 需要 `unrar` 后端程序（随附的 `assets/unrar.exe` 已内置）。

## 使用说明

```
BookerPdfTool -h    # 查看帮助
BookerPdfTool -v    # 查看版本
```

帮助中也列出了全部子命令和参数。

### fm-office：Office 转 PDF

将 DOC/XLS/PPT 转换为 PDF（Windows）。

```
BookerPdfTool fm-office <fname>
```

| 参数 | 说明 |
| --- | --- |
| `fname` | 文件或目录名 |

支持 `doc/docx/xls/xlsx/ppt/pptx`；传入目录时批量转换目录下的所有文件。

### comp：压缩 PDF

对 PDF 内嵌图片做压缩处理后写回，并清理无用数据。

```
BookerPdfTool comp <fname>
```

| 参数 | 说明 |
| --- | --- |
| `fname` | PDF 文件名 |

### ext：抽取 PDF 图片

将 PDF 每一页的内嵌图片抽出来保存为 PNG（或整页截图，需 `-w`）。

```
BookerPdfTool ext <fname> [-d 输出目录] [-w]
```

| 参数 | 说明 |
| --- | --- |
| `fname` | 文件或目录名 |
| `-d, --dir` | 输出目录，默认当前目录 |
| `-w, --whole` | 整页截图，不抽取内嵌图片 |

### 2html：PDF 转 HTML

将 PDF 每页渲染为 HTML，并识别行内代码 / 代码块 / 标题、处理缩进。

```
BookerPdfTool 2html <fname> [-d 输出目录]
```

| 参数 | 说明 |
| --- | --- |
| `fname` | 文件或目录名 |
| `-d, --dir` | 输出目录，默认当前目录 |

### anime4k-auto：Anime4K 增强图片

用 Anime4K 对图片做缩放增强，适合扫描版图片降噪提清晰度。

```
BookerPdfTool anime4k-auto <fname> [-G] [-t 线程数]
```

| 参数 | 说明 |
| --- | --- |
| `fname` | 文件或目录名 |
| `-G, --gpu` | 使用 GPU 推理 |
| `-t, --threads` | 线程数，默认 8 |

需要将 Anime4KCPP_CLI 加入系统 PATH（见「可选依赖」）。

### pack：图片打包成 PDF

把指定目录下的图片按顺序打包为 PDF；配合正则可以把按关键词命名的一组图片分别打包。

```
BookerPdfTool pack <dir> [-r 正则] [--jb2]
```

| 参数 | 说明 |
| --- | --- |
| `dir` | 图片所在目录 |
| `-r, --regex` | 按文件名匹配关键词分组，每组生成一个 PDF |
| `--jb2` | 生成 JBIG2 编码的 PDF，体积更小 |

### tog-bw：黑白反转纠正

自动检测反色的黑白扫描图（平均灰度低于阈值的图片）并反转，纠正「黑底白字」。

```
BookerPdfTool tog-bw <fname> [-t 线程数] [-s 阈值]
```

| 参数 | 说明 |
| --- | --- |
| `fname` | 文件或目录名 |
| `-t, --threads` | 线程数，默认 8 |
| `-s, --thres` | 平均灰度低于该值才视为反色并反转，默认 50 |

### crack-zip：破解加密压缩包

对加密的 ZIP / RAR 压缩包进行字典爆破，成功后解压到同名目录。

```
BookerPdfTool crack-zip <fname> [-p 密码字典] [-t 线程数]
```

| 参数 | 说明 |
| --- | --- |
| `fname` | ZIP / RAR 文件名 |
| `-p, --pw` | 密码字典文件，默认使用随附的 `assets/PwdDic.txt` |
| `-t, --threads` | 线程数，默认 8 |

### auto：一键全自动处理 PDF

一条命令完成整套流程：`ext` 抽图 → `tog-bw` 黑白纠正 → `anime4k-auto` 增强 → 二值化 → `pack` 打包回 PDF。

```
BookerPdfTool auto <fname> [-t 线程数] [-G] [-w]
```

| 参数 | 说明 |
| --- | --- |
| `fname` | PDF 文件或目录名 |
| `-t, --threads` | 线程数，默认 8 |
| `-G, --gpu` | Anime4K 使用 GPU |
| `-w, --whole` | 抽取整页截图 |

原文件会备份为 `xxx.pdf.bak`。同样需要 Anime4KCPP_CLI。

### pick-scan：甄别扫描版 PDF

根据页面图片占比判断每本 PDF 是「扫描版」还是「文字版」，并移动到对应子目录。

```
BookerPdfTool pick-scan <dir> [-i 图片面积占比] [-s 扫描页占比] [-t 线程数]
```

| 参数 | 说明 |
| --- | --- |
| `dir` | 存放 PDF 的目录 |
| `-i, --imgs-area-rate` | 页面图片面积占比高于该值则视为扫描页，默认 0.8 |
| `-s, --scanned-pg-rate` | 扫描页占比高于该值则整本视为扫描版，默认 0.8 |
| `-t, --threads` | 线程数，默认 8 |

结果分别移动到 `dir/扫描版` 与 `dir/文字版`。

### pdg2pic：PDG 转图片

按封面、卷首、正文等前缀排序，把 PDG 文件复制为顺序编号的 PNG。

```
BookerPdfTool pdg2pic <dir> [-o 输出目录]
```

| 参数 | 说明 |
| --- | --- |
| `dir` | PDG 文件所在目录 |
| `-o, --output-dir` | 输出目录，默认与输入目录相同 |

### dedup：PDF 去重

基于文本 n-gram 哈希的相似度判断，把重复的 PDF 移到 `dup` 目录，并把历史向量保存到数据库文件，供后续批次比对。

```
BookerPdfTool dedup <fname> [-t 相似度阈值] [--db 历史库文件]
```

| 参数 | 说明 |
| --- | --- |
| `fname` | PDF 文件或目录名 |
| `-t, --thres` | 相似度阈值，高于该值判定重复，默认 0.9 |
| `--db` | 保存历史向量的 `.npy` 文件，默认 `临时目录/dedup.npy` |

## 协议

本项目基于 SATA 协议发布。

您有义务为此开源项目点赞，并考虑额外给予作者适当的奖励。

## 赞助我们

![](https://home.apachecn.org/img/about/donate.jpg)

## 另见

+   [ApacheCN 学习资源](https://docs.apachecn.org/)
+   [计算机电子书](http://it-ebooks.flygon.net)
+   [布客新知](http://flygon.net/ixinzhi/)

CHANGELOG.md
# 历史记录

v2023.7.19

+   打包 PDF 支持 JBIG2 模式（仅限 Windows）

v2023.4.5

+   新增抓取 Medium 文章列表功能

v2023.3.30

+   移除只用于 Windows 的依赖

v 2023.2.27

+   `fetch-links`功能增加 JSON 模式（`-J`）

v2023.2.23

+   增加`batch-link`功能

v2023.2.20

+   再增加 RAR 密码破解功能

v2023.2.18

+   引入 GIT 高级初始化`git-init`，逐个文件提交`git-commit`，逐个提交推送`git-push`功能

v2023.2.11

+   新增破解 ZIP 密码的功能
+   引入`7z`，`unrar`，`jbig2`，`Pdg2Pic`实用工具，通过`util.asset()`获取路径。

v2023.2.9

+   新增筛选扫描版 PDF 的功能

v2023.2.7

+   修复 EPUB 目录提取的若干 BUG

v2023.2.5

+   添加一键智能处理 PDF 功能

v2023.2.2

+   添加 anime4k 图像优化工具（需要`Anime4KCPP_CLI`）
+   添加转换 CDRIVE 日志功能


v2023.1.31

+   添加 ChatGPT 命令行工具

v2023.1.30

+   修复 CMYK 不能保存为 PNG 的错误

v2013.1.29

+  增加 C 风格代码排版功能

v2023.1.21

+  修复并优化 EPUB 目录展示和章节提取功能

v2023.1.20

+   新增 DOC 转换为 PDF 功能，`ppt2pdf`命令改为`office2pdf`
+   新增 PDF 压缩功能

v2023.1.19

+   新增根据公众号导出元数据抓取公众号文章的功能

v2023.1.2

+   优化 TOMD 功能

v2022.12.20

+   HTML 转 MD 支持多线程

v2022.12.11

+   修复获取文章链接功能的BUG

v2022.11.29

+   修复图片截断无法加载的问题

v2022.11.24

+   增加 Waifu2x 自动处理图片功能
    需要将`waifu2x-caffe.exe`所在目录添加到系统变量 PATH 中
+   删除分拣图片功能

v2022.11.14

+   增加查看 EPUB 目录功能

v2022.11.6

+   增加 PDF 打包功能
+   增加展开目录功能

v2022.11.4

+   增加 PDF 提取图片功能
+   增加分拣图片功能

v2022.10.20

+   增加 PPT 转 PDF 功能
+   移除 BILIBILI 下载功能

v2022.8.15

+   增加转换 MD、压缩 EPUB、关键帧提取、BILIBILI 下载功能

v2022.7.2

+   增加字数统计功能，与 Word 规则一致

v2022.6.28

+   增加修复文件名的功能

v2022.4.23

+   建立项目

BookerPdfTool/__main__.py
#!/usr/bin/env python3.7
# -*- coding: utf-8 -*-

import argparse
from . import __version__
from .util import *
from .pdf_tool import *
from .zip_tool import *
from .toggle_bw import *
    
def main():
    parser = argparse.ArgumentParser(prog="BookerPdfTool", description="iBooker PDF tool", formatter_class=argparse.RawDescriptionHelpFormatter)
    parser.add_argument("-v", "--version", action="version", version=f"BookerPdfTool version: {__version__}")
    parser.set_defaults(func=lambda x: parser.print_help())
    subparsers = parser.add_subparsers()
    
   
    office2pdf_parser = subparsers.add_parser("fm-office", help="doc/xls/ppt to pdf")
    office2pdf_parser.add_argument("fname", help="file name")
    office2pdf_parser.set_defaults(func=office2pdf_handle)
    
    comp_pdf_parser = subparsers.add_parser("comp", help="compress pdf")
    comp_pdf_parser.add_argument("fname", help="file name")
    comp_pdf_parser.set_defaults(func=comp_pdf)


    ext_pdf_parser = subparsers.add_parser("ext", help="extract odf into images")
    ext_pdf_parser.add_argument("fname", help="file name")
    ext_pdf_parser.add_argument("-d", "--dir", default='.', help="path to save")
    ext_pdf_parser.add_argument("-w", "--whole", action='store_true', default=False, help="whether to clip the whole page")
    ext_pdf_parser.set_defaults(func=ext_pdf)

    pdf2html_parser = subparsers.add_parser("2html", help="convert pdf page to html")
    pdf2html_parser.add_argument("fname", help="file name")
    pdf2html_parser.add_argument("-d", "--dir", default='.', help="path to save")
    pdf2html_parser.set_defaults(func=pdf2html)

    anime4k_auto_parser = subparsers.add_parser("anime4k-auto", help="process imgs with waifu2x")
    anime4k_auto_parser.add_argument("fname", help="file or dir name")
    anime4k_auto_parser.add_argument("-G", "--gpu", action='store_true', help="whether to use GPU")
    anime4k_auto_parser.add_argument("-t", "--threads", help="num of threads", type=int, default=8)
    anime4k_auto_parser.set_defaults(func=anime4k_auto_handle)

    pack_pdf_parser = subparsers.add_parser("pack", help="package images into pdf")
    pack_pdf_parser.add_argument("dir", help="dir name")
    pack_pdf_parser.add_argument("-r", "--regex", help="regex of keyword for grouping")
    pack_pdf_parser.add_argument("--jb2", action='store_true', help="rwhether to generate jb2 encoding pdf")
    pack_pdf_parser.set_defaults(func=pack_pdf)


    toggle_bw_parser = subparsers.add_parser("tog-bw", help="check if image colors reversed and then toggle them")
    toggle_bw_parser.add_argument("fname", help="file or dir name")
    toggle_bw_parser.add_argument("-t", "--threads", type=int, default=8, help="num of thread")
    toggle_bw_parser.add_argument("-s", "--thres", type=int, default=50, help="threshold less than which the color will be regarded as black")
    toggle_bw_parser.set_defaults(func=toggle_bw_handle)

    ck_zip_parser = subparsers.add_parser("crack-zip", help="crack encrypted zip")
    ck_zip_parser.add_argument("fname", help="ZIP fname")
    ck_zip_parser.add_argument("-p", "--pw", default=asset('PwdDic.txt'), help="password dict")
    ck_zip_parser.add_argument("-t", "--threads", type=int, default=8, help="num of threads")
    ck_zip_parser.set_defaults(func=crack_zip)


    pdf_auto_parser = subparsers.add_parser("auto", help="auto process pdf")
    pdf_auto_parser.add_argument("fname", help="pdf fname or dirname")
    pdf_auto_parser.add_argument("-t", "--threads", type=int, default=8, help="num of threads")
    pdf_auto_parser.add_argument("-G", "--gpu", action='store_true', help="whether to use GPU")
    pdf_auto_parser.add_argument("-w", "--whole", action='store_true', default=False, help="whether to clip the whole page")
    pdf_auto_parser.set_defaults(func=pdf_auto_handle)

    pick_scan_parser = subparsers.add_parser("pick-scan", help="pick scanned pdf")
    pick_scan_parser.add_argument("dir", help="dirname of pdfs")
    pick_scan_parser.add_argument("-i", "--imgs-area-rate", type=float, default=0.8, help="rate of imgs area in page area, above which a page will be regarded as scanned")
    pick_scan_parser.add_argument("-s", "--scanned-pg-rate", type=float, default=0.8, help="rate of scanned pages in whole doc, above which a pdf will be regarded as scanned")
    pick_scan_parser.add_argument("-t", "--threads", type=int, default=8, help="num of threads")
    pick_scan_parser.set_defaults(func=pick_scanned_pdf)


    args = parser.parse_args()
    args.func(args)

if __name__ == "__main__": main()